core/services/run_queue.go

```diff
@@ -34,81 +36,188 @@ type RunQueue interface {
 	}
 }
 
 // Start prepares the job runner for accepting runs to execute.
 func (rq *runQueue) Start() error {
+	go rq.orchestrateWorkers()
 	return nil
 }
 
 // Stop closes all open worker channels.
 func (rq *runQueue) Stop() {
-	rq.workersMutex.Lock()
-	rq.stopRequested = true
-	rq.workersMutex.Unlock()
-	rq.workersWg.Wait()
+	close(rq.chStop)
+	<-rq.chDone
 }
 
 // Run tells the job runner to start executing a job
 func (rq *runQueue) Run(run *models.JobRun) {
-	rq.workersMutex.Lock()
-	if rq.stopRequested {
-		rq.workersMutex.Unlock()
-		return
+	select {
+	case rq.chRuns <- run:
+	case <-rq.chStop:
 	}
+}
 
-	runID := run.ID.String()
-	defer numberRunsQueued.Inc()
-	if queueCount, present := rq.workers[runID]; present {
-		rq.workers[runID] = queueCount + 1
-		rq.workersMutex.Unlock()
-		return
-	}
-	rq.workers[runID] = 1
-	numberRunQueueWorkers.Set(float64(len(rq.workers)))
-	rq.workersMutex.Unlock()
+func (rq *runQueue) orchestrateWorkers() {
+	defer close(rq.chDone)
+	for {
+		select {
+		case run := <-rq.chRuns:
+			worker, exists := rq.workers[*run.JobSpecID]
+			if !exists {
+				worker = &singleJobSpecWorker{make(chan *models.JobRun), make(chan struct{}), make(chan struct{})}
+				rq.workers[*run.JobSpecID] = worker
+				atomic.AddInt32(&rq.numWorkers, 1)
+				go rq.runSingleJobSpecWorker(*run.JobSpecID, worker)
+				worker.chAddRun <- run
+			} else {
+				select {
+				case worker.chAddRun <- run:
+				case <-worker.chDone:
+					// If the worker is just spinning down as new work is coming in, allow it
+					// to die and spin up a new one.
+					worker = &singleJobSpecWorker{make(chan *models.JobRun), make(chan struct{}), make(chan struct{})}
+					rq.workers[*run.JobSpecID] = worker
+					atomic.AddInt32(&rq.numWorkers, 1)
+					go rq.runSingleJobSpecWorker(*run.JobSpecID, worker)
+					worker.chAddRun <- run
+				}
+			}
 
-	rq.workersWg.Add(1)
-	go func() {
-		for {
-			rq.workersMutex.Lock()
-			queueCount := rq.workers[runID]
-			if queueCount <= 0 {
-				delete(rq.workers, runID)
-				numberRunQueueWorkers.Set(float64(len(rq.workers)))
-				rq.workersMutex.Unlock()
-				break
+		case jobSpecID := <-rq.chWorkerDone:
+			delete(rq.workers, jobSpecID)
+			if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+				panic("numWorkers should never be < 0")
 			}
-			rq.workers[runID] = queueCount - 1
-			rq.workersMutex.Unlock()
 
-			if err := rq.runExecutor.Execute(run.ID); err != nil {
-				logger.Errorw(fmt.Sprint("Error executing run ", runID), "error", err)
+		case <-rq.chStop:
+			for _, worker := range rq.workers {
+				worker.chStop <- struct{}{}
+				<-worker.chDone
+				if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+					panic("numWorkers should never be < 0")
+				}
 			}
+			return
 		}
+	}
+}
 
-		rq.workersWg.Done()
+func (rq *runQueue) runSingleJobSpecWorker(jobSpecID models.ID, worker *singleJobSpecWorker) {
+	defer close(worker.chDone)
+
+	promNumberRunQueueWorkersStarted.Inc()
+
+	var (
```

Comment:
Lot of shared state here, I feel like this entire method could be turned into an object with some methods which would make it a lot easier to test some of these edge cases.